doc/_extensions/vuln_table.py

```diff
@@ -0,0 +1,133 @@
+        for key in keys:
+            entry = nodes.entry()
+            entry += nodes.paragraph(text=key)
+            row += entry
+        thead.append(row)
+
+        # table body
+        rows = []
+        for vuln in cache["body"]:
+            row = nodes.row()
+            rows.append(row)
+
+            for key in keys:
+                entry = nodes.entry()
+                para = nodes.paragraph(text=vuln[key])
+                entry += para
+                row += entry
+
+        tbody = nodes.tbody()
+        tbody.extend(rows)
+        tgroup += tbody
+
+        return [table]
+
+def progress(response: PipelineResponse) -> None:
+    """Download progress callback."""
+
+    mb_tot = response.context["data_stream_total"] / (1024 * 1024)
+    mb_curr = response.context["download_stream_current"] / (1024 * 1024)
+    pct = int((mb_curr / mb_tot) * 100)
+    s = term_width_line(
+        bold("Downloading vulnerability files... ")
+        + f"{mb_curr:.2f} of {mb_tot:.2f} Mb [{pct} %]"
+    )
+    logger.info(s, nonl=pct < 100)
+
+def download_vuln_table(app: Sphinx) -> None:
+    """Download vulnerability table from cache"""
+
+    # Check if local cache exists
+    try:
+        with open(Path(app.doctreedir) / Path("vuln_cache.pickle"), "rb") as cache_file:
+            local_cache = pickle.load(cache_file)
+    except FileNotFoundError:
+        logger.info("No vulnerability cache found locally")
+        local_cache = None
+
+    # Check internet connection
+    try:
+        requests.get("https://ncsdocsa.blob.core.windows.net", timeout=3)
+    except (requests.ConnectionError, requests.Timeout):
+        logger.info("Could not retrieve vulnerability table online")
+        if local_cache:
+            logger.info("Using local vulnerability cache")
+            app.env.vuln_cache = local_cache
+        return
+
+    cc = ContainerClient.from_connection_string(AZ_CONN_STR_PUBLIC, AZ_CONTAINER)
+    remote_files = sorted(
+        [b for b in cc.list_blobs()],
+        key=lambda b : b.last_modified,
+        reverse=True)
+
+    target = remote_files[0]
+    if local_cache and local_cache["md5"] == target.content_settings["content_md5"]:
+        app.env.vuln_cache = local_cache
+        logger.info("Up to date vulnerability table found in cache")
+        return
+
+    bc = cc.get_blob_client(target)
+    res = bc.download_blob(raw_response_hook=progress).content_as_text()
+    app.env.vuln_cache = json.loads(res)
+    app.env.vuln_cache["md5"] = target.content_settings["content_md5"]
+
+    with open(Path(app.doctreedir) / Path("vuln_cache.pickle"), "wb") as cache_file:
+        pickle.dump(app.env.vuln_cache, cache_file)
+    logger.info("Vulnerability table cached locally")
+
+def setup(app: Sphinx):
+    app.add_directive("vuln-table", VulnTable)
+
+    app.connect("builder-inited", download_vuln_table)
+
+    return {
+        'version': '0.1',
+        'parallel_read_safe': True,
+        'parallel_write_safe': True,
```

Comment:
nit: use `"` or `'`, avoid mixing both. I'd suggest to use the `black` autoformatter.